source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+#include "common/event/scaled_range_timer.h"
+
+#include <chrono>
+#include <cmath>
+#include <memory>
+
+#include "envoy/event/timer.h"
+
+#include "common/common/assert.h"
+#include "common/common/scope_tracker.h"
+
+namespace Envoy {
+namespace Event {
+
+/**
+ * Implementation of RangeTimer that can be scaled by the backing manager object.
+ *
+ * Instances of this class exist in one of 4 states:
+ *  - inactive: not enabled
+ *  - pending: enabled, min timeout not elapsed
+ *  - active: enabled, min timeout elapsed, max timeout not elapsed
+ */
+class ScaledRangeTimerManager::ScaledRangeTimer final : public RangeTimer {
+public:
+  ScaledRangeTimer(TimerCb callback, ScaledRangeTimerManager& manager)
+      : manager_(manager), callback_(callback),
+        pending_timer_(manager.dispatcher_.createTimer([this] { onPendingTimerComplete(); })) {}
+
+  ~ScaledRangeTimer() override { disableTimer(); }
+
+  void disableTimer() override {
+    struct Dispatch {
+      Dispatch(ScaledRangeTimer& timer) : timer(timer) {}
+      ScaledRangeTimer& timer;
+      void operator()(const Inactive&) {}
+      void operator()(const Pending&) { timer.pending_timer_->disableTimer(); }
+      void operator()(const Active& active) {
+        timer.manager_.disableActiveTimer(active.duration, active.bucket_position);
+      }
+    };
+    absl::visit(Dispatch(*this), state_);
+    state_.emplace<Inactive>();
+  }
+
+  void enableTimer(const std::chrono::milliseconds& min_ms, const std::chrono::milliseconds& max_ms,
+                   const ScopeTrackedObject* scope) override {
+    disableTimer();
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
```

Comment:
Use of manager_.dispatcher_.timeSource().monotonicTime() doesn't seem as correct as getting the cached monotonic time from libevent, but that may be a fix to timeSource instead of an issue with your code.

We should get that fixed.

Reply:
I've switched to using Dispatcher::approximateMonotonicTime, which I think should be sufficient.